### PR TITLE
Add unit testing framework for synapse

### PR DESCRIPTION
**Pull request by sajithaliyanage**
## Purpose
Provide a built-in way to execute a unit test against a given synapse configuration artifact such as sequence, proxy, api, local entry, endpoint.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: modules/commons/src/main/java/org/apache/synapse/commons/emulator/RequestProcessor.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        } catch (Exception e) {
+            logger.error("Error while parsing xmlString to DOM", e);
+        }
+        return null;
+    }
+
+    private static String nodeToString(Node node) {
+        StringWriter sw = new StringWriter();
+        try {
+            Transformer t = TransformerFactory.newInstance().newTransformer();
```
Comment (milindaperera) [inline modules/commons/src/main/java/org/apache/synapse/commons/emulator/RequestProcessor.java]: single letter variable name here. Let's update name with some meaning full name

File: modules/core/src/main/java/org/apache/synapse/unittest/TestingAgent.java [hunk 2]
```diff
@@ -0,0 +1,396 @@
+/*
+ Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (milindaperera) [inline modules/core/src/main/java/org/apache/synapse/unittest/TestingAgent.java]: small formatting issue here
let's fix the same in other locations as well

File: modules/commons/src/main/java/org/apache/synapse/commons/emulator/RequestProcessor.java (hunk 1 above)
Comment (sajithaliyanage) [reply modules/commons/src/main/java/org/apache/synapse/commons/emulator/RequestProcessor.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/1452

File: modules/core/src/main/java/org/apache/synapse/unittest/TestingAgent.java (hunk 2 above)
Comment (sajithaliyanage) [reply modules/core/src/main/java/org/apache/synapse/unittest/TestingAgent.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/1452

**Comment by sajithaliyanage** [comment]
License header issue fixed in https://github.com/wso2/wso2-synapse/pull/1452

File: modules/core/src/main/java/org/apache/synapse/core/axis2/Axis2SynapseEnvironment.java [hunk 3]
```diff
@@ -217,6 +222,37 @@ public void setSeverDebugMode(ServerContextInformation contextInformation) {
         }
     }
 
+    /**
+     * This method is to set the unit test mode is enabled.
+     * unit test message context and environment initializes
```
Comment (sdkottegoda) [inline modules/core/src/main/java/org/apache/synapse/core/axis2/Axis2SynapseEnvironment.java]: Shall we correct the javadoc?

File: modules/core/src/main/java/org/apache/synapse/unittest/ConfigurationDeployer.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+     * @return response of the artifact deployment and the synapse configuration as a Pair<>
+     */
+    Pair<SynapseConfiguration, String> deployApiArtifact(OMElement inputElement, String fileName)
+            throws AxisFault {
+
+        //create new API deployer object
+        APIDeployer apiResourceDeployer = new APIDeployer();
+
+        //create a synapse configuration and set all axis2 configuration to it
+        SynapseConfiguration synapseConfiguration = UnitTestingExecutor.getExecuteInstance().getSynapseConfiguration();
```
Comment (sdkottegoda) [inline modules/core/src/main/java/org/apache/synapse/unittest/ConfigurationDeployer.java]: Shall we extract the duplicate code segments to a separate method?

File: modules/core/src/main/java/org/apache/synapse/unittest/MockServiceCreator.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            HttpConsumerContext emulator = Emulator.getHttpEmulator()
+                    .consumer()
+                    .host(host)
+                    .port(port)
+                    .context(context);
+
+            for (ServiceResource resource : resources) {
+                routeThroughResourceMethod(resource, emulator);
+            }
+
```
Comment (sdkottegoda) [inline modules/core/src/main/java/org/apache/synapse/unittest/MockServiceCreator.java]: Let's remove unnecessary new lines

Comment (sajithaliyanage) [reply modules/core/src/main/java/org/apache/synapse/unittest/MockServiceCreator.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/1452

File: modules/core/src/main/java/org/apache/synapse/unittest/ConfigurationDeployer.java (hunk 4 above)
Comment (sajithaliyanage) [reply modules/core/src/main/java/org/apache/synapse/unittest/ConfigurationDeployer.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/1452

File: modules/core/src/main/java/org/apache/synapse/core/axis2/Axis2SynapseEnvironment.java (hunk 3 above)
Comment (sajithaliyanage) [reply modules/core/src/main/java/org/apache/synapse/core/axis2/Axis2SynapseEnvironment.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/1452

**Pull request merged**